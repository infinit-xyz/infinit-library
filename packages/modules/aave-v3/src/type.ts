import { Address } from 'viem'

export type AaveV3Registry = {
  poolAddressProviderRegistry?: Address
  poolAddressesProvider?: Address
  liquidationLogic?: Address
  supplyLogic?: Address
  eModeLogic?: Address
  flashLoanLogic?: Address
  borrowLogic?: Address
  bridgeLogic?: Address
  poolLogic?: Address
  configuratorLogic?: Address
  reservesSetupHelper?: Address
  poolProxy?: Address
  lendingPools?: Record<string, LendingPool>
  walletBalanceProvider?: Address
  uiIncentiveDataProvider?: Address
  uiPoolDataProviderV3?: Address
  aaveProtocolDataProvider?: Address
  l2PoolImpl?: Address
  poolConfiguratorImpl?: Address
  aclManager?: Address
  aaveOracle?: Address
  aaveEcosystemReserveV2Proxy?: Address
  aaveEcosystemReserveController?: Address
  aaveEcosystemReserveV2Impl?: Address
  reserveInterestRateStrategies?: Record<string, Address>
  wrappedTokenGatewayV3?: Address
  l2Encoder?: Address
  aTokenImpl?: Address
  delegationAwareATokenImpl?: Address
  stableDebtTokenImpl?: Address
  variableDebtTokenImpl?: Address
  poolConfiguratorProxy?: Address
  rewardsControllerImpl?: Address
  rewardsControllerProxy?: Address
  pullRewardsTransferStrategy?: Address
  emissionManager?: Address
  aggregatorApi3Adapters?: Record<string, Address>
  aggregatorPythAdapters?: Record<string, Address>
  aggregatorBandAdapters?: Record<string, Address>
}

export type LendingPool = {
  underlyingToken: Address
  interestRateStrategy: Address
  aToken: Address
  stableDebtToken: Address
  variableDebtToken: Address
}

export type Modify<T, R> = Omit<T, keyof R> & R
