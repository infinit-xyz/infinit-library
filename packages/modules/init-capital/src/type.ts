import { Address } from 'viem'

export type InitCapitalRegistry = {
  proxyAdmin?: Address
  accessControlManager?: Address
  api3ProxyOracleReaderProxy?: Address
  initOracleProxy?: Address
  configProxy?: Address
  liqIncentiveCalculatorProxy?: Address
  posManagerProxy?: Address
  initCoreProxy?: Address
  riskManagerProxy?: Address
  lsdApi3ProxyOracleReaderProxy?: Address
  pythOracleReaderProxy?: Address
  moneyMarketHookProxy?: Address
  initLens?: Address
  api3ProxyOracleReaderImpl?: Address
  initOracleImpl?: Address
  configImpl?: Address
  liqIncentiveCalculatorImpl?: Address
  posManagerImpl?: Address
  initCoreImpl?: Address
  riskManagerImpl?: Address
  lsdApi3ProxyOracleReaderImpl?: Address
  pythOracleReaderImpl?: Address
  moneyMarketHookImpl?: Address
  lendingPoolImpl?: Address
  irms?: Record<string, Address>
  lendingPools?: Record<string, LendingPools>
}

export type LendingPools = {
  underlyingToken: Address
  lendingPool: Address
  irm: Address
}
