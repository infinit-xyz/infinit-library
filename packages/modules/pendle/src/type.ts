import { Address } from 'viem'

export type PendleV3Registry = {
  pendleRouterV4?: Address
  routerStorageV4?: Address
  actionAddRemoveLiqV3?: Address
  actionCallbackV3?: Address
  actionMiscV3?: Address
  actionSimple?: Address
  actionSwapPTV3?: Address
  actionSwapYTV3?: Address
  pendleRouterStatic?: Address
  actionStorageStatic?: Address
  actionInfoStatic?: Address
  actionMarketAuxStatic?: Address
  actionMarketCoreStatic?: Address
  actionMintRedeemStatic?: Address
  actionVePendleStatic?: Address
  pendleLimitRouter?: Address
  pendleLimitRouterProxy?: Address
  proxyAdmin?: Address
  baseSplitCodeFactoryContract?: Address
  oracleLib?: Address
  pendleGaugeControllerMainchainUpg?: Address
  pendleMarketFactoryV3?: Address
  pendlePYLpOracle?: Address
  pendleSwap?: Address
  pendleYieldContractFactory?: Address
  votingEscrowPendleMainchain?: Address
  pendleMsgSendEndpointUpgImpl?: Address
  pendleMsgSendEndpointUpgProxy?: Address
  pendleVotingContollerUpgImpl?: Address
  pendleVotingControllerUpgProxy?: Address
}
