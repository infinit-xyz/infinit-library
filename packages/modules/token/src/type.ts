import { Address, Hex } from 'viem'

export enum TokenType {
  'InfinitERC20' = 'InfinitERC20',
  'InfinitERC20Burnable' = 'InfinitERC20Burnable',
}

export type Token = { type: TokenType }
export type Proxy = { proxyAddress: Address; implementation: Address }

export type TokenRegistry = {
  tokens?: Record<Address, Token>
  tokenDistributors?: Address[]
  accumulativeMerkleDistributors?: Record<Address, Proxy>
  merkleTree?: {
    root: string
    merkle: Record<Address, { amount: string; proof: Hex[] }>
  }
}
