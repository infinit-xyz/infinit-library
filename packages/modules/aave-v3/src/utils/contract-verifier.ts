import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { AaveV3Registry, LendingPool } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class AaveV3ContractVerifier extends BaseContractVerifier<AaveV3Registry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: AaveV3Registry): Promise<ContractInfo[]> {
    const {
      lendingPools,
      reserveInterestRateStrategies,
      poolProxy,
      poolConfiguratorProxy,
      rewardsControllerProxy,
      api3Adapters,
      pythAdapters,
      bandAdapters,
      ...contractToAddressRegistry
    } = registry

    let contracts: ContractInfo[] = []

    for (const address of Object.values(contractToAddressRegistry)) {
      contracts.push({ address })
    }

    if (registry.poolAddressesProvider) {
      if (poolProxy) {
        contracts.push({
          address: poolProxy,
          constructorArgs: [registry.poolAddressesProvider],
        })
      }
      if (poolConfiguratorProxy) {
        contracts.push({
          address: poolConfiguratorProxy,
          constructorArgs: [registry.poolAddressesProvider],
        })
      }
    }

    if (lendingPools && registry.poolConfiguratorProxy) {
      const lendingPoolContractInfos = await this.getLendingPoolContractInfos(registry.poolConfiguratorProxy, lendingPools)
      contracts = contracts.concat(lendingPoolContractInfos)
    }

    if (reserveInterestRateStrategies) {
      const reserveInterestRateStrategyContractInfos = await this.getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies)
      contracts = contracts.concat(reserveInterestRateStrategyContractInfos)
    }

    if (api3Adapters) {
      const api3AdaptersContractInfos = await this.getAdapterContractInfos(api3Adapters)
      contracts = contracts.concat(api3AdaptersContractInfos)
    }

    if (pythAdapters) {
      const pythAdaptersContractInfos = await this.getAdapterContractInfos(pythAdapters)
      contracts = contracts.concat(pythAdaptersContractInfos)
    }

    if (bandAdapters) {
      const bandAdaptersContractInfos = await this.getAdapterContractInfos(bandAdapters)
      contracts = contracts.concat(bandAdaptersContractInfos)
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }

  private async getLendingPoolContractInfos(poolConfiguratorProxy: Address, lendingPools: Record<string, LendingPool>) {
    const contracts: ContractInfo[] = []
    for (const info of Object.values(lendingPools)) {
      for (const [contractName, address] of Object.entries(info)) {
        if (contractName === 'underlyingToken') continue
        contracts.push({
          address,
          constructorArgs: contractName !== 'interestRateStrategy' ? [poolConfiguratorProxy] : undefined,
        })
      }
    }
    return contracts
  }

  private async getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies: Record<string, Address>) {
    const contracts: ContractInfo[] = []
    for (const address of Object.values(reserveInterestRateStrategies)) {
      contracts.push({ address })
    }
    return contracts
  }

  private async getAdapterContractInfos(adapters: Record<string, Address>) {
    const contracts: ContractInfo[] = []
    for (const address of Object.values(adapters)) {
      contracts.push({ address })
    }
    return contracts
  }
}
