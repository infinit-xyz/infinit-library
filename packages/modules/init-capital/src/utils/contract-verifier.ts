import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { InitCapitalRegistry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class TokenContractVerifier extends BaseContractVerifier<InitCapitalRegistry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: InitCapitalRegistry): Promise<ContractInfo[]> {
    const { lendingPools, irms, ...contractToAddressRegistry } = registry
    const contracts: ContractInfo[] = []

    for (const address of Object.values(contractToAddressRegistry)) {
      contracts.push({ address })
    }

    if (lendingPools) {
      for (const { lendingPool, irm } of Object.values(lendingPools)) {
        contracts.push({ address: lendingPool })
        contracts.push({ address: irm })
      }
    }

    if (irms) {
      for (const address of Object.values(irms)) {
        contracts.push({ address })
      }
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
