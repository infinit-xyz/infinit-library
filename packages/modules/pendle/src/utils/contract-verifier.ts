import { PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { PendleRegistry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class TokenContractVerifier extends BaseContractVerifier<PendleRegistry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(_registry: PendleRegistry): Promise<ContractInfo[]> {
    const contracts: ContractInfo[] = []

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
