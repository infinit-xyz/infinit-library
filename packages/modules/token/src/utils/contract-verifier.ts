import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { TokenRegistry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class TokenContractVerifier extends BaseContractVerifier<TokenRegistry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: TokenRegistry): Promise<ContractInfo[]> {
    const { accumulativeMerkleDistributors, tokens } = registry
    const contracts: ContractInfo[] = []

    if (tokens) {
      for (const address of Object.keys(tokens)) {
        contracts.push({ address: address as Address })
      }
    }
    if (accumulativeMerkleDistributors) {
      for (const [proxyAddress, { implementation }] of Object.entries(accumulativeMerkleDistributors)) {
        contracts.push({ address: proxyAddress as Address })
        contracts.push({ address: implementation })
      }
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
