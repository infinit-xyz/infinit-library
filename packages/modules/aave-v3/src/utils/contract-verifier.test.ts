import { describe, expect, test } from 'vitest'

import { createPublicClient, http } from 'viem'
import { holesky } from 'viem/chains'

import { AaveV3Registry } from '../type'
import { AaveV3ContractVerifier } from './contract-verifier'

// Create a subclass to expose the protected method for testing
class TestAaveV3ContractVerifier extends AaveV3ContractVerifier {
  public async testGetContracts(registry: AaveV3Registry) {
    return this.getContracts(registry)
  }
}

describe('AaveV3ContractVerifier.ts', () => {
  const registry: AaveV3Registry = {
    poolAddressProviderRegistry: '0xed4b96e65132287cdfceb5bb891b12c4f35a7d07',
    supplyLogic: '0xf4d4555bccb25f57d38402295fc4cb22aa1f1e7b',
    borrowLogic: '0x9df952e8f2f752d65f383987104145adadbdae74',
    liquidationLogic: '0xe360332c6a6a611cfded42daebaebe075cccda03',
    eModeLogic: '0x76765b6546cb530009a07abc7b9884db399243a8',
    bridgeLogic: '0x3e1643471239f86f173850720ffff63f5e33838e',
    configuratorLogic: '0x5a4d5610ba08d9a019aa7f68b5317996072c5630',
    poolLogic: '0xd6d69359ef662106dc767c35cfd818a6a536de16',
    aaveEcosystemReserveV2Proxy: '0xaa9f6f6a281f176968574ff2fea55783fdd56b16',
    aaveEcosystemReserveController: '0x1b476c1a75e3d61ffcc8162070e42b32c97b42fb',
    aaveEcosystemReserveV2Impl: '0x7a2a9a30d942305e7d52b8e309b8b717d8be15d7',
    poolAddressesProvider: '0x44a7818e44e2e328210a7ca773a5ca4044fedf26',
    reservesSetupHelper: '0x7390e9eafc6a6afe332e2f5ac8e53e677dc5755a',
    walletBalanceProvider: '0xf9de6d863d2474537e077620c86fba79a67809a5',
    uiIncentiveDataProvider: '0xd3c7d30d88b00e14e8debb6c8d2351311147bd1d',
    uiPoolDataProviderV3: '0x4d43bb9a1bf9dc564250ccbc70d277b87af1daa5',
    emissionManager: '0x8dbcb843d6763655ca7addeae1cc03ef723f3e33',
    aaveProtocolDataProvider: '0x3d3b400d21e518471b5929db5b20de7b33454ac6',
    flashLoanLogic: '0x8f679eaf072d6a346f2a1a4597eb5e22337ed5ee',
    poolConfiguratorImpl: '0xb6d8fa4a79e402a12a91d93806f0778e9a6d5b38',
    aclManager: '0x33ebc17143e38ed877514f16f37e4ac124528fcb',
    aaveOracle: '0x00e9faa434f07aeaa14db2d3cbaa65fbfd2397a2',
    rewardsControllerImpl: '0x312578528c108700265819b89b8337557f1b3890',
    reserveInterestRateStrategies: {
      rateStrategyVolatileOne: '0x67e2eb9e8c38b2597e835a3914822f92bb43a193',
    },
    l2PoolImpl: '0x5b0c1fda2b61427fc8e84b1e3810f3847a63c3e2',
    poolProxy: '0xF1E061CC318160B8490D7E9E9Ba13b1FaA927f2b',
    poolConfiguratorProxy: '0xF2a93eD27f61969bD7EE8F170Bb4024a481Fb6D5',
    rewardsControllerProxy: '0xA3C641B31a4aDA7C4676954E8F4E35158352c778',
    wrappedTokenGatewayV3: '0xe8bc6291e6794631bf683e4c757d2212e2944009',
    l2Encoder: '0x0dd3adc925cfd041661be3083281000b59cff101',
    aTokenImpl: '0x07fc7e7d7908f4ccd5604c45702ad92867498424',
    delegationAwareATokenImpl: '0xc302fbb999d7e36d2326ccf4c663d9893b06a278',
    stableDebtTokenImpl: '0x35eaaa18aa6ed5e0c17b51b030076d4476cb6c01',
    variableDebtTokenImpl: '0x6bf2168cebd62eb0c89f5d88ed8ff1fa41c24577',
    pullRewardsTransferStrategy: '0x554d2af481073b3552ced42d12032c9c9057481f',
    lendingPools: {
      WETH: {
        underlyingToken: '0x6B5817E7091BC0C747741E96820b0199388245EA',
        interestRateStrategy: '0x67e2eb9e8c38b2597e835a3914822f92bb43a193',
        aToken: '0xC1972a55235cCD222b1FAd713025652374cf31F1',
        stableDebtToken: '0x2A556b9e34dc204aa7Ce924342f544351Ed74885',
        variableDebtToken: '0x0bA89BEA0EB039B907Fa5d7CFb63D9044529Cd14',
      },
    },
  }

  test('getContracts: should match snapshot', async () => {
    const verifier = new TestAaveV3ContractVerifier(
      await createPublicClient({
        chain: holesky,
        transport: http('https://1rpc.io/holesky'),
      }),
      {
        apiKey: process.env.HOLESKY_ETHERSCAN_API_KEY ?? '',
        apiUrl: 'https://api-holesky.etherscan.io/api',
        url: 'https://holesky.etherscan.io',
      },
    )
    expect(verifier.testGetContracts(registry)).resolves.toMatchSnapshot()
  })
})
