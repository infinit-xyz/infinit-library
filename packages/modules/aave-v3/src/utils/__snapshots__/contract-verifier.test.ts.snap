// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AaveV3ContractVerifier.ts > getContracts: should match snapshot 1`] = `
[
  {
    "address": "0xed4b96e65132287cdfceb5bb891b12c4f35a7d07",
  },
  {
    "address": "0xf4d4555bccb25f57d38402295fc4cb22aa1f1e7b",
  },
  {
    "address": "0x9df952e8f2f752d65f383987104145adadbdae74",
  },
  {
    "address": "0xe360332c6a6a611cfded42daebaebe075cccda03",
  },
  {
    "address": "0x76765b6546cb530009a07abc7b9884db399243a8",
  },
  {
    "address": "0x3e1643471239f86f173850720ffff63f5e33838e",
  },
  {
    "address": "0x5a4d5610ba08d9a019aa7f68b5317996072c5630",
  },
  {
    "address": "0xd6d69359ef662106dc767c35cfd818a6a536de16",
  },
  {
    "address": "0xaa9f6f6a281f176968574ff2fea55783fdd56b16",
  },
  {
    "address": "0x1b476c1a75e3d61ffcc8162070e42b32c97b42fb",
  },
  {
    "address": "0x7a2a9a30d942305e7d52b8e309b8b717d8be15d7",
  },
  {
    "address": "0x44a7818e44e2e328210a7ca773a5ca4044fedf26",
  },
  {
    "address": "0x7390e9eafc6a6afe332e2f5ac8e53e677dc5755a",
  },
  {
    "address": "0xf9de6d863d2474537e077620c86fba79a67809a5",
  },
  {
    "address": "0xd3c7d30d88b00e14e8debb6c8d2351311147bd1d",
  },
  {
    "address": "0x4d43bb9a1bf9dc564250ccbc70d277b87af1daa5",
  },
  {
    "address": "0x8dbcb843d6763655ca7addeae1cc03ef723f3e33",
  },
  {
    "address": "0x3d3b400d21e518471b5929db5b20de7b33454ac6",
  },
  {
    "address": "0x8f679eaf072d6a346f2a1a4597eb5e22337ed5ee",
  },
  {
    "address": "0xb6d8fa4a79e402a12a91d93806f0778e9a6d5b38",
  },
  {
    "address": "0x33ebc17143e38ed877514f16f37e4ac124528fcb",
  },
  {
    "address": "0x00e9faa434f07aeaa14db2d3cbaa65fbfd2397a2",
  },
  {
    "address": "0x312578528c108700265819b89b8337557f1b3890",
  },
  {
    "address": "0x5b0c1fda2b61427fc8e84b1e3810f3847a63c3e2",
  },
  {
    "address": "0xe8bc6291e6794631bf683e4c757d2212e2944009",
  },
  {
    "address": "0x0dd3adc925cfd041661be3083281000b59cff101",
  },
  {
    "address": "0x07fc7e7d7908f4ccd5604c45702ad92867498424",
  },
  {
    "address": "0xc302fbb999d7e36d2326ccf4c663d9893b06a278",
  },
  {
    "address": "0x35eaaa18aa6ed5e0c17b51b030076d4476cb6c01",
  },
  {
    "address": "0x6bf2168cebd62eb0c89f5d88ed8ff1fa41c24577",
  },
  {
    "address": "0x554d2af481073b3552ced42d12032c9c9057481f",
  },
  {
    "address": "0xF1E061CC318160B8490D7E9E9Ba13b1FaA927f2b",
    "constructorArgs": [
      "0x44a7818e44e2e328210a7ca773a5ca4044fedf26",
    ],
  },
  {
    "address": "0xF2a93eD27f61969bD7EE8F170Bb4024a481Fb6D5",
    "constructorArgs": [
      "0x44a7818e44e2e328210a7ca773a5ca4044fedf26",
    ],
  },
  {
    "address": "0x67e2eb9e8c38b2597e835a3914822f92bb43a193",
    "constructorArgs": undefined,
  },
  {
    "address": "0xC1972a55235cCD222b1FAd713025652374cf31F1",
    "constructorArgs": [
      "0xF2a93eD27f61969bD7EE8F170Bb4024a481Fb6D5",
    ],
  },
  {
    "address": "0x2A556b9e34dc204aa7Ce924342f544351Ed74885",
    "constructorArgs": [
      "0xF2a93eD27f61969bD7EE8F170Bb4024a481Fb6D5",
    ],
  },
  {
    "address": "0x0bA89BEA0EB039B907Fa5d7CFb63D9044529Cd14",
    "constructorArgs": [
      "0xF2a93eD27f61969bD7EE8F170Bb4024a481Fb6D5",
    ],
  },
  {
    "address": "0x67e2eb9e8c38b2597e835a3914822f92bb43a193",
  },
]
`;
